### PR TITLE
[installer]: create readme

### LINKED ISSUE #6704: [installer]: create a readme
A readme is needed to provide some context for the Installer. This should give an overview on how to use it, some common pitfalls and links to relevant issues that may provide external users with problems/questions

### PR DESCRIPTION
**Pull request by MrSimonEmms**
## Description

Create an initial readme for the Installer. This is designed to provide an overview to the technical usage of the application.

## Related Issue(s)

Fixes #6704

## How to test

Read it and follow the instructions

## Release Notes

```release-note
[Installer]: create readme
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** after the PR has been reviewed.

Associated issue: *#6704*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: installer/README.md [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+VERSION="${VERSION:-main.1844}" # Build ID taken from https://werft.gitpod-dev.com/
+
+docker run -it --rm \
+  -v $PWD:/gitpod \
+  -v $HOME/.kube:/root/.kube \
+  eu.gcr.io/gitpod-core-dev/build/installer:${VERSION} \
+  $@
+```
+
+Save this in your `$PATH` as `gitpod-installer` and make executable
```
Comment (kylos101) [inline installer/README.md:40]: Where do folks normally put SH scripts for product? I feel like `$HOME/.gitpod/bin` isn't the right place..but there a better sub-folder to use than bin? I ask because I'm a long time Windows user now using Ubuntu. Can you think of any other products that persist scripts to user machines?

If we had it in bin (or whatever), we could do something like:

Add `gitpod-installer.sh` to your PATH with:

```bash
export PATH=$PATH:$HOME/.gitpod/bin
```

File: installer/README.md [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+```shell
+gitpod-installer init > config.yaml
+```
+
+## Validate
+
+```shell
+# Checks the validity of the configuration YAML
+gitpod-installer validate config --config /gitpod/config.yaml
```
Comment (kylos101) [inline installer/README.md:52]: After the `validate config` command, I think it would be good to say this right away.

> Any errors here must be fixed before deploying. See [Config](#config) for more details.

File: installer/README.md [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+# Checks that your cluster is ready to install Gitpod
+gitpod-installer validate cluster --kubeconfig /root/.kube/config --config /gitpod/config.yaml
+```
+
+Any errors here must be fixed before deploying. See [Cluster Dependencies](#cluster-dependencies)
```
Comment (kylos101) [inline installer/README.md:58]: If we do the config above, we can then just do the cluster here like:

> Any errors here must be fixed before deploying. See [Cluster Dependencies](#cluster-dependencies)
> for more details.

File: installer/README.md [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+Save this in your `$PATH` as `gitpod-installer` and make executable
+
+## Generate the base config
+
+```shell
+gitpod-installer init > config.yaml
+```
+
+## Validate
```
Comment (kylos101) [inline installer/README.md:48]: Should config.yaml be changed to  ./gitpod.config.yaml?

Also, before the Validate step, it may (may?) be good to say something like...

> We assume at this point you might make customizations - please do. Here's a link to the configuration details. After your done, please validate the configuration before proceeding.

File: installer/README.md [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+The Installer generates a ConfigMap with the metadata of every Kubernetes
+object generated by the Installer. This can be retrieved to remove Gitpod
+from your cluster.
+
+```shell
+kubectl get configmaps gitpod-app -o jsonpath='{.data.app\.yaml}' \
+  | kubectl delete -f - # Piping to this will delete automatically
+```
+
+**Important**. This may leave certain objects still in your Kubernetes
```
Comment (kylos101) [inline installer/README.md:87]: Yes, good call on mentioning this. :100:

File: installer/README.md [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+
+> Not every parameter is discussed in this table, just ones that are likely
+> to need changing. The full config structure is available in [config.go](/installer/pkg/config/v1/config.go).
+
+| Property | Required | Description | Notes |
+| --- | --- | --- | --- |
+| `domain` | Y | The domain to deploy to | This will need to be changed on every deployment |
+| `kind` | Y | Installation type to run - for most users, this will be `Full` | Can be `Full`, `Meta` or `Workspace` |
+| `metadata.region` | Y | Location for your `objectStorage` provider | If using Minio, set to `local` |
+| `workspace.runtime.containerdRuntimeDir` | Y | The location of containerd on host machine | Common values are: <ul><li>`/run/containerd/io.containerd.runtime.v2.task/k8s.io` (K3s)</li><li>`/var/lib/containerd/io.containerd.runtime.v2.task/k8s.io` (AWS/GCP)</li><li>`/run/containerd/io.containerd.runtime.v1.linux/k8s.io` (Azure)</li><li>`/run/containerd/io.containerd.runtime.v1.linux/moby`</li></ul> |
```
Comment (kylos101) [inline installer/README.md:113]: :rock: :star:

File: installer/README.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+5. `kubectl apply`
+
+# Quickstart
+
+## Create a gitpod-installer script
+
+```shell
+#!/bin/bash
+
+VERSION="${VERSION:-main.1844}" # Build ID taken from https://werft.gitpod-dev.com/
```
[the remark this answers is not in the document]
Comment (kylos101) [reply installer/README.md:31]: I see you opened #6766, good man @MrSimonEmms ! :+1: :+1:

**Review by kylos101 — COMMENTED** [review]
Hey there @MrSimonEmms, Happy Friday! 

:handshake: :beer: :bee:

File: installer/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:31]: @kylos101 I'm planning on writing an RFC for Gitpodders next week to discuss the options. My personal favourite is to do a regular monthly release on-or-around the 7th of the month (then I do a release on my birthday) and to number them as Ubuntu does - eg, this month's would be 21.11, next would be 21.12. These would be then treated as `stable` and anything in any branch would be treated as `edge` (I'm easy on terminology) - FYI, I don't think we have the capacity to offer an LTS version yet.

The one advantage to a regular monthly cadence is that the community can then expect things to drop on a certain date. 

I think what I've done in the readme is good enough for today, but will need to be figured out in the next few weeks.

File: installer/README.md [hunk 8]
```diff
@@ -0,0 +1,312 @@
+# Installer
+
+The best way to get started with Gitpod self-hosted
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:3]: Yeah, I thought I'd make it explicit - then we all know it (you could argue that "self-hosted" is redundant, since we're dogfooding it)

File: installer/README.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+to entry for new users - the base config was many hundreds of lines long
+and did not have effective validation on it. By contrast, the Installer's
+config is < 50 lines long and can be fully validated before running.
+
+Conceptually, the process is now:
+1. generate a base config
+2. amend the config for your own use-case
+3. validate
+4. render the Kubernetes YAML
+5. `kubectl apply`
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:22]: Done

File: installer/README.md (hunk 1 above)
Comment (MrSimonEmms) [reply installer/README.md:40]: This is one of those things that I think you could end up writing a whole readme on. As the plan is to change this to the [actual binary release](#6766) rather than a Bash script to wrap the Docker image, I sort of think we can keep this simple and assume an amount of knowledge from the reader.

Ultimately, my ambition is to be able to have something like how Helm does it with [binary](https://helm.sh/docs/intro/install/#from-the-binary-releases) and [script](https://helm.sh/docs/intro/install/#from-script)

File: installer/README.md (hunk 4 above)
Comment (MrSimonEmms) [reply installer/README.md:48]: Done

File: installer/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:3]: In fact, I am getting rid of "self-hosted". We're using it for SaaS, so it's the best way<full stop>

File: installer/README.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+> refer to the [Gitpod docs](https://www.gitpod.io/docs/self-hosted/latest/installation)
+> for instructions on how to create the infrastructure.
+
+# Key Concepts
+
+The Installer is a complete replacement for our Helm charts. Over time,
+this had grown to be too complex to effectively support and was a barrier
+to entry for new users - the base config was many hundreds of lines long
+and did not have effective validation on it. By contrast, the Installer's
+config is < 50 lines long and can be fully validated before running.
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:15]: I'd quite like to keep this paragraph as an answer to "why are we doing it?". When I announced it on Discord a few days ago, there were a few "why are you moving away from Helm, it's the de facto standard and everyone knows it" questions - perfectly reasonable questions (I asked the same).

That said, I think you're right about the positioning of it. I've made a change - WDYT?

File: installer/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kylos101) [reply installer/README.md:38]: Agreed on all counts. :smile: 

Btw, RFC is a great idea. It's a tricky subject.

File: installer/README.md (hunk 1 above)
Comment (kylos101) [reply installer/README.md:47]: Nice, thank you for sharing. Actual binaries would rock, and, I like those helm examples - they seem like the right fit for where things are presently.

File: installer/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (kylos101) [reply installer/README.md]: Creating a FAQ section and moving this there was a great idea!!! It reads well, too.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** after the PR has been reviewed.

Associated issue: *#6704*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** after the PR has been reviewed.

Associated issue: *#6704*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: installer/README.md [hunk 11]
```diff
@@ -0,0 +1,441 @@
...
+5. `kubectl apply`
+
+# Quickstart
+
+## Build your binary
+
+> There is work to be done on improving the distribution of the binary
+> [#6766](https://github.com/gitpod-io/gitpod/issues/6766)
+
+From [werft](https://werft.gitpod-dev.com), find the tag you want, such as
```
Comment (princerachit) [inline installer/README.md:39]: maybe reword this to `search your tag`?

Comment (MrSimonEmms) [reply installer/README.md]: Done

**Review by princerachit — COMMENTED** [review]
PTAL

**Comment by princerachit** [comment]
I will suggest you to paste the output of each command here or a screenshot in the comments of this PR

File: installer/README.md [hunk 12]
```diff
@@ -0,0 +1,441 @@
...
+# Add the versions in
+objcopy --add-section versionManifest=./versions.yaml ./installer
+```
+
+You can also `--debug-version-file` if you want to use `go run`
+
+## Generate the base config
+
+```shell
+gitpod-installer init > gitpod.config.yaml
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:56]: No, we're using the global path name

File: installer/README.md [hunk 13]
```diff
@@ -0,0 +1,441 @@
...
+# Checks the validity of the configuration YAML
+gitpod-installer validate config --config /gitpod/gitpod.config.yaml
+```
+
+Any errors here must be fixed before deploying. See [Config](#config) for
+more details.
+
+```shell
+# Checks that your cluster is ready to install Gitpod
+gitpod-installer validate cluster --kubeconfig /root/.kube/config --config /gitpod/gitpod.config.yaml
```
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:79]: Good spot on the abs paths - I'll switch to the standard `/path/to/file`

File: installer/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (MrSimonEmms) [reply installer/README.md:56]: Actually, I think you're right - I'll change to `./installer`

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 420244a66575da82218000aa842b7b1c3a5c60e7</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6767#issuecomment-987882462" title="LGTM">csweichel</a>*

Associated issue: *#6704*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
